### PR TITLE
Support multiple data backends for Truffle DB's workspace

### PR DESCRIPTION
**Pull request by gnidan**
See commits, but this PR significantly refactors the Workspace class to support multiple adapters, as specified by constructor argument to `new Workspace()`. Most notably, this PR includes two adapters: the existing `jsondown` filesystem adapter, as well as a new memory-based adapter.

Summary of changes:

- Stub a bunch of additional schema types.
- Use `sources` as ID field for Compilation
- Add ID fields to definitions for each collection
- Rename `fetchAll` method to `all`
- Make pouch module a directory
- Define useful types
- Implement schema-generic interface to PouchDB
- Implement generic Databases adapter for jsondown
- Realize generic Databases types for DataModel
- Move Workspace class out of pouch module
- Decouple separate concerns in Workspace
- Document type parameter name conventions
- Delegate Workspace adapter instantiation by name
- Add memory-based adapter for Workspace
- Update persistence tests for dual-adapter world
- Update default fs directory to .db/json/

### CODE REVIEW

File: packages/truffle-db/src/workspace/pouch/types.ts [hunk 1]
```diff
@@ -11,6 +20,149 @@ export type MutationMapping<C, M> = {
...
+    this.definitions = definitions;
+
+    PouchDB.plugin(PouchDBDebug);
+    PouchDB.plugin(PouchDBFind);
+
+    this.collections = Object.keys(definitions)
+      .map((resource: CollectionName<C>) => ({
+        [resource]: this.createDatabase(resource)
+      }))
+      .reduce((a, b) => ({ ...a, ...b })) as CollectionDatabases<C>;
```
Comment (cds-amal) [inline packages/truffle-db/src/workspace/pouch/types.ts]: Is the initial value missing here?

Comment (gnidan) [reply packages/truffle-db/src/workspace/pouch/types.ts]: Oops, you're right!

File: packages/truffle-db/src/workspace/pouch/memory.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import PouchDB from "pouchdb";
+import PouchDBMemoryAdapter from "pouchdb-adapter-memory";
+
+import {
+  Databases,
+  MutationMapping,
+  ResourceMapping,
+  WorkspaceDatabasesOptions
```
Comment (cds-amal) [inline packages/truffle-db/src/workspace/pouch/memory.ts]: This doesn't seem to be used

**Comment by gnidan** [comment]
Thanks @fainashalts!

**Comment by gnidan** [comment]
Also thanks to @cds-amal for being a sounding board while writing this :)

**Pull request merged**